### ISSUE TITLE
What happens when there's no AP in an annotation dictionary?

### ISSUE THREAD

**Issue by MPBailey** [pdf-association/pdf-issues#23]
In Table 166 AP is not required in a couple of cases, but immediately after the table the text says "A PDF reader shall render the appearance dictionary without regard to any other keys and values in the annotation dictionary and shall ignore the values of the C, IC, Border, BS, BE, BM, CA, ca, H, DA, Q, DS, LE, LL, LLE, and Sy keys."

So what should the renderer do if there is no AP?

For a zero-size annotation I recommend that it doesn't render anything, and maybe the requirement that the border is drawn completely inside the annotation rectangle (12.5.4 para 1) is enough to state that anyway.

I think for Popup and Projection annotations the correct behaviour is stated adequately elsewhere, so the lack of any allowance for a missing AP here simply muddies the water. It's less clear that there is a clear statement for Link annotations anywhere. Did I miss something?

A good start would be to amend the para after Table 166 to start "For all annotations containing an AP entry, a PDF reader ...", and then amending the following note to say "Requiring an appearance dictionary for **most** annotation**s** ..."

**Comment by petervwyatt** [comment]
Some comments:

- Table 166 possibly already defines what you call a "zero-size annotation": "Annotations where the value of the **Rect** key consists of an array where the value at index 1 is equal to the value at index 3 and the value at index 2 is equal to the value at index 4. NOTE (2020) The bullet point above was changed from “or” to “and” in this document to match requirements in other published ISO PDF standards (such as PDF/A)." - however the choice of "AND" means that a zero-wdith-only or zero-height-only annot would still paint a pixel thin line (thanks to the scan conversion rules). 

- I agree that the description wording choice in Table 166 de-emphasises the newly stated importance of **AP** in PDF 2.0 and changing to something like "_(Usually required, PDF 1.2)_" would be much better.

- for the para after Table 166 I do not think "For all annotations containing an AP entry, a PDF reader ..." is what is meant. To me, the current wording says that PDF readers shall _always_ ignore all the appearance description keys and thus shall not try and simulate any appearance streams. ISO 32K is written from a compliance point of view so since AP shall always be present excepting for the few cases stated then that is what shall be used. However a PDF _writer_ will need to use those keys to create or regenerate AP appearance streams using their own implementation-dependent algorithms.

**Comment by lrosenthol** [comment]
> PDF readers shall always ignore all the appearance description keys and thus shall not try and simulate any appearance streams

Correct.

> However a PDF writer will need to use those keys to create or regenerate AP appearance streams using their own implementation-dependent algorithms

Correct.

**Comment by MPBailey** [comment]
Thanks, so we have concensus on what a renderer should do, but what we don't yet have is some text that says that. So I'm going to suggest:
* Amending the AP entry to say "usually required", as discussed above.
* Amending the 1st para after the table to: “_For all annotations containing an AP entry,_ a PDF reader shall render the appearance dictionary without regard to any other keys and values in the annotation dictionary_. If no AP entry is present the annotation shall not be rendered. When rendering the annotation a PDF reader_ shall ignore the values of the C, IC, Border, BS, BE, BM, CA, ca, H, DA, Q, DS, LE, LL, LLE, and Sy keys.”

The "when rendering an annotation" bit is because a reader that is doing something other than rendering should NOT ignore those keys, of course.

**Comment by lrosenthol** [comment]
Isn't the point of `BM, CA and ca` to specify how to composite the appearance onto the page?   If that's true, then those keys (if present) *shall* be used for rendering the AP/annotation.

**Comment by petervwyatt** [comment]
@lrosenthol you are correct for `BM`, but not for `ca` or `CA` as they are both currently explicitly specified to be the content stream and exclude themselves for when AP streams are present. 

But having **BM** including in the list of keys in the 1st para after Table 166 is incorrect! 

Quoting Table 166:

- **BM**: (_Optional; PDF 2.0_) The blend mode that shall be used when painting the annotation onto the page ...

- **ca** : (_Optional; PDF 2.0_) When regenerating the annotation's appearance stream, this is the opacity value (11.2, "Overview of transparency") that shall be used for all nonstroking operations on all visible elements of the annotation in its closed state (including its background and border) but not the popup window that appears when the annotation is opened.
Default value: 1.0
The specified value shall not be used if the annotation has an appearance stream (see 12.5.5, "Appearance streams"); in that case, the appearance stream shall specify any transparency.
If no explicit appearance stream is defined for the annotation, and the processor is not able to regenerate the appearance, the annotation may be painted by implementation-dependent means that do not necessarily conform to the PDF imaging model; in this case, the effect of this entry is implementation-dependent as well.

- **CA**: (_Optional; PDF 1.4, PDF 2.0 for non-markup annotations_) When regenerating the annotation's appearance stream, this is the opacity value (11.2, "Overview of transparency") that shall be used for stroking all visible elements of the annotation in its closed state, including its background and border, but not the popup window that appears when the annotation is opened.
If a ca entry is not present in this dictionary, then the value of this CA entry shall also be used for nonstroking operations as well. Default Value: 1.0
The specified value shall not be used if the annotation has an appearance stream (12.5.5, "Appearance streams"); in that case, the appearance stream shall specify any transparency.
If no explicit appearance stream is defined for the annotation, and the processor is not able to regenerate the appearance, the annotation may be painted by implementation-dependent means that do not necessarily conform to the PDF imaging model; in this case, the effect of this entry is implementation-dependent as well

**Comment by petervwyatt** [comment]
Once #42 and #94 are resolved, I believe the issues here will also be resolved.

Review those first and then see if anything is outstanding (and noting that PDF is silent about the rules of annot appearance stream regeneration which is a far more significant issue).

**Comment by petervwyatt** [comment]
PDF TWG to review next time after reviewing PDF/A, PDF/X requirements.

**Comment by MPBailey** [comment]
I've pasted the relevant text from PDF/X-6 (15930-9) below.

This says nothing about what must be done when there is no AP key, and I suspect that Note 1 implies an expectation at the time that we wrote this that AP (or appearance subdictionaries for widgets) would be required in all cases.

I don't think this constrains what we do for this issue at all.

**6.12.3 Annotation appearances**
For all annotation dictionaries containing an AP key, the appearance dictionary that it defines as its
value shall contain only the N key. If an annotation dictionary’s Subtype key has a value of Widget and
its FT key has a value of Btn, the value of the N key shall be an appearance subdictionary otherwise the
value of the N key shall be an appearance stream.
NOTE 1 Appearance dictionaries are required to be present in a file and rendered appropriately according to
ISO 32000-2:2020, Table 166 and the following paragraph. This ensures the reliable rendering of the annotations.
NOTE 2 In accordance with the requirements of ISO 32000-2:2020, 12.7.5.2.3 and 12.7.5.2.4 a Button form
field will have multiple appearance states, each one associated with the specific values that the button can take.
